tests/gordo_components/builder/test_mlflow_utils.py

```diff
@@ -0,0 +1,406 @@
+    "x,expected",
+    [
+        (datetime(1970, 1, 1), 0),
+        (datetime(1970, 1, 1, 0, 0, 1), 1000),
+        (datetime(1970, 1, 1, 0, 0, 0, 1000), 1),
+    ],
+)
+def test_datetime_to_ms_since_epoch(x, expected):
+    """
+    Test that datetime is correctly converted to ms since Unix epoch
+    """
+    assert mlu._datetime_to_ms_since_epoch(x) == expected
+
+
+def test_get_batch_kwargs(metadata):
+    """
+    Test that dicts are correctly converted to MLflow types or errors raised
+    """
+
+    def _test_convert_metadata(metadata):
+        batch_kwargs = mlu.get_batch_kwargs(metadata)
+
+        assert all(type(m) == Metric for m in batch_kwargs["metrics"])
+        assert all(type(p) == Param for p in batch_kwargs["params"])
+
+    # With cross validation and metric scores
+    _test_convert_metadata(metadata)
+
+    # With cross validation, no scores
+    metadata["model"]["cross-validation"].pop("scores")
+    _test_convert_metadata(metadata)
+
+    # no cross validation or scores
+    metadata["model"].pop("cross-validation")
+    _test_convert_metadata(metadata)
+
+
+@pytest.mark.parametrize(
+    "secret_str,keys,keys_valid",
+    [
+        ("dummy1:dummy2:dummy3", ["key1", "key2", "key3"], True),
+        ("dummy1:dummy2:dummy3", ["key1", "key2"], False),
+    ],
+)
+def test_get_kwargs_from_secret(monkeypatch, secret_str, keys, keys_valid):
+    """
+    Test that service principal kwargs are generated correctly if env var present
+    """
+    env_var_name = "TEST_SECRET"
+
+    # TEST_SECRET doesn't exist as env var
+    with pytest.raises(ValueError):
+        mlu.get_kwargs_from_secret(env_var_name, keys)
+
+    # TEST_SECRET exists as env var
+    monkeypatch.setenv(name=env_var_name, value=secret_str)
+    if keys_valid:
+        kwargs = mlu.get_kwargs_from_secret(env_var_name, keys)
+        for key, value in zip(keys, secret_str.split(":")):
+            assert kwargs[key] == value
+    else:
+        with pytest.raises(ValueError):
+            mlu.get_kwargs_from_secret(env_var_name, keys)
+
+
+def test_workspace_spauth_kwargs():
+    """Make sure an error is thrown when env vars not set"""
+    with pytest.raises(ValueError):
+        mlu.get_workspace_kwargs()
+
+    with pytest.raises(ValueError):
+        mlu.get_spauth_kwargs()
+
+
+@mock.patch("gordo_components.builder.mlflow_utils.MlflowClient", autospec=True)
+def test_mlflow_context_log_metadata(MockClient, tmp_dir, recwarn, metadata):
+    """
+    Test that call to wrapped function initiates MLflow logging or throws warning
+    """
+
+    mlflow.set_tracking_uri(f"file:{tmp_dir}")
+
+    mock_client = MockClient()
+    mock_client.log_batch.return_value = "test"
+
+    # Function with a metadata dict returned
+    with mlu.mlflow_context("returns metadata", "unique_key", {}, {}) as (
+        mlflow_client,
+        run_id,
+    ):
+        mlu.log_metadata(mlflow_client, run_id, metadata)
+
+    assert mock_client.log_batch.called
```

Comment:
I hope none of these tests ever break.. :sweat_smile: I definitely don't want to try fixing them.

Reply:
No way, these tests are like :tropical_drink: :sunglasses: :palm_tree: .